### PR TITLE
fix: add recursive type detection for the go sdk

### LINKED ISSUE #5871: Zenith: custom return types require at least one method defined
Custom function return types require at least one empty method defined on them to be detected by codegen (note the JSON tags are currently required, #5860).

```go
// This is for a module "foo"

package main

import (
	"context"
)

type Foo struct{}

type X struct {
	Message string `json:"message"`
}

// This function definition is required, or we don't detect and incorporate this into schema generation.
// func (x *X) Void() {}

func (m *Foo) MyFunction(ctx context.Context, stringArg string) X {
	return X{Message: stringArg}
}
```

```
$ echo '{foo{myFunction(stringArg: "hello"){message}}' | dagger query -m ".
Error: failed to get loaded module ID: input:1: host.directory.asModule.serve failed to install module schema: schema validation failed: input:1813: Undefined type X.
```

### PR DESCRIPTION
**Pull request by jedevc**
This fixes #5871.

For example, previously this would have failed, since the `Void` method was required.

```go
// foo is the module

type Foo struct{}

type X struct {
	Message string `json:"message"`
}

// This function definition is required, or we don't detect and
// incorporate this into schema generation.
// func (x *X) Void() {}

func (m *Foo) MyFunction(ctx context.Context, stringArg string) X {
	return X{Message: stringArg}
}
```

similarly with this example:

```go
// foo is the module

type Foo struct {
	Example X
}

type X struct {
	Message string `json:"message"`
}
```

With this patch, we attempt to recursively investigate function signatures, structs, etc, to try and find additional types we need to incorporate into the graphql structure defintion.

### CODE REVIEW

**Comment by vito** [comment]
Nice!

I think we have enough of a framework in place that a test for this would be somewhat straightforward to add.

Something like this:

https://github.com/dagger/dagger/blob/bec193e06806b46275f0df7d5584ff7b8a91cdb3/core/integration/module_test.go#L491-L509

Lately I've started to favor just directly embedding Go code in the tests since it's a bit of a pain to maintain dozens of tiny testdata files:

https://github.com/dagger/dagger/blob/bec193e06806b46275f0df7d5584ff7b8a91cdb3/core/integration/module_test.go#L558-L563

**Comment by jedevc** [comment]
Aha, awesome - I'll work on a test on Monday, I'm wrapping up for the week now :)

**Comment by jedevc** [comment]
@vito I've pushed a test - it's currently failing though, until https://github.com/dagger/dagger/blob/268c1062a902c8bade88eedde8daee5ed90affdf/sdk/go/modules/sdks.go#L22 is updated though.

Ideally the tests should work with the in-tree version of these deps, not sure if you have ideas for how we could startup dev versions of these?

**Comment by vito** [comment]
@jedevc Until #5900 lands, the expected workflow is to run the following to build + push a new image:

```sh
../../hack/dev ../../bin/dagger run go run ./cmd/bootstrap/ --publish vito/dagger-sdk-go:(git rev-parse HEAD)
```

**Review by sipsma — APPROVED** [review]
one nit, otherwise LGTM!

File: sdk/go/codegen/generator/go/templates/modules.go [hunk 1]
```diff
@@ -107,6 +109,50 @@ func (funcs goTemplateFuncs) moduleMainSrc() string {
...
 		// Add the object to the module
 		createMod = dotLine(createMod, "WithObject").Call(Add(Line(), objType))
+		added[obj.Name()] = struct{}{}
+	}
+
+	// do another pass for all the other nested types that we might encounter
+	for len(ps.extraTypes) > 0 {
+		tps := ps.extraTypes
+		ps.extraTypes = nil
+		for _, tp := range tps {
```
[the remark this answers is not in the document]
Comment (jedevc) [reply sdk/go/codegen/generator/go/templates/modules.go]: Yup, this is pretty possible to do - we just pre-process the objects into types, and then do the iteration based on that.

**Pull request merged**